dnf-json

```diff
@@ -200,64 +250,129 @@ class DnfJsonRequestHandler(BaseHTTPRequestHandler):
+                        "DepsolveError",
+                        (
+                            "There was a problem depsolving "
+                            f"{arguments['package-specs']}: {e}"
+                        )
+                    )
+                except dnf.exceptions.Error as e:
+                    self.response_with_dnf_error(
+                        type(e).__name__,
+                        f"Error occurred when setting up repo: {e}")
+        finally:
+            cache = None
+            with process_count.get_lock():
+                process_count.value -= 1
+                # Only the last process alive is making the cleaning.
+                # This is because we don't want to delete directories that might
+                # be used by a live dnf instance.
+                if process_count.value == 0:
+                    self.clean_cache()
+                else:
+                    # If the process is not the last, it sends it's used folders
+                    # to the last one.
+                    self.send_cache_folder_used()
+
+    def send_cache_folder_used(self):
+        # make a list of the folder used this round and send them to the queue
+        for cache_folder in self.cache_folders:
+            cache_folders_queue.put(cache_folder)
+
+    def clean_cache(self):
+        log.info("clean_cache")
+        # receive all the cache folders used on the other processes.
+        try:
+            while(True):
+                self.cache_folders.append(cache_folders_queue.get_nowait())
+                log.debug("received %s", self.cache_folders[-1])
+        except queue.Empty as e:
+            pass
+        # load the old save from previous run
+        cache_state = self.load_cache_state_from_disk()
+        # update the caches utilisation from the list of used folders
+        cache_state.update_used(self.cache_folders)
+        # clean all the folders unused since a timeout duration
+        cache_state.clean_unused(self.cachedir, CACHE_TIMEOUT)
+        # store cache state for next iteration
+        self.store_cache_state_to_disk(cache_state)
+
+    def load_cache_state_from_disk(self):
+        try:
+            with open(f"{self.cachedir}/cache_state.pkl", "rb") as inp:
+                return pickle.load(inp)
+        except FileNotFoundError as e:
+            return CacheState()
+
+    def store_cache_state_to_disk(self, cache_state):
+        with open(f"{self.cachedir}/cache_state.pkl", "wb") as out:
+            pickle.dump(cache_state, out, pickle.HIGHEST_PROTOCOL)
 
 log.info("Starting the dnf-json server")
 
 LISTEN_FDS = int(os.environ.get("LISTEN_FDS", 0))
 SOCK_PATH = "/run/osbuild-dnf-json/"
 SOCK_NAME = "api.sock"
 
+# The dnf-json web server has to use forks to serve the requests. Because the
+# dnf library is leaking memory in its Cpp side.
 class SystemDActivationSocketServer(socketserver.ForkingMixIn, socketserver.UnixStreamServer):
```

Comment:
Hey! Thanks for the awesome review! So we unfortunately have to use forks here with `ForkingMixIn` Because using threads leads to loss of ram each time a request is answered. To answer the question, yes `ForkingMixIn` uses forks.